src/test/java/org/apache/sysds/test/functions/federated/datagen/FederatedRandTest.java

```diff
@@ -0,0 +1,156 @@
+	}
+
+	@Parameterized.Parameters
+	public static Collection<Object[]> data() {
+		return Arrays.asList(new Object[][] {
+			{12, 10, 0, 1, 0.7, 123},
+			{1200, 10, -10, 10, 0.9, 123}
+		});
+	}
+
+	@Test
+	public void federatedRand1CP() {
+		federatedRand(TEST_NAME1, Types.ExecMode.SINGLE_NODE);
+	}
+
+	@Test
+	public void federatedRand2CP() {
+		federatedRand(TEST_NAME2, Types.ExecMode.SINGLE_NODE);
+	}
+
+	@Test
+	public void federatedRand3CP() {
+		federatedRand(TEST_NAME3, Types.ExecMode.SINGLE_NODE);
+	}
+
+	public void federatedRand(String TEST_NAME, Types.ExecMode execMode) {
+		boolean sparkConfigOld = DMLScript.USE_LOCAL_SPARK_CONFIG;
+		Types.ExecMode platformOld = rtplatform;
+		rtplatform = execMode;
+		if(rtplatform == Types.ExecMode.SPARK) {
+			DMLScript.USE_LOCAL_SPARK_CONFIG = true;
+		}
+
+		getAndLoadTestConfiguration(TEST_NAME);
+		String HOME = SCRIPT_DIR + TEST_DIR;
+
+		// write input matrices
+		int halfRows = rows / 2;
+		// We have two matrices handled by a single federated worker
+		double[][] X1 = getRandomMatrix(halfRows, cols, 0, 1, 1, 42);
+		double[][] X2 = getRandomMatrix(halfRows, cols, 0, 1, 1, 1340);
+
+		writeInputMatrixWithMTD("X1", X1, false, new MatrixCharacteristics(halfRows, cols, blocksize, halfRows * cols));
+		writeInputMatrixWithMTD("X2", X2, false, new MatrixCharacteristics(halfRows, cols, blocksize, halfRows * cols));
+
+		int port1 = getRandomAvailablePort();
+		int port2 = getRandomAvailablePort();
+		Thread t1 = startLocalFedWorkerThread(port1, FED_WORKER_WAIT_S);
+		Thread t2 = startLocalFedWorkerThread(port2);
+
+		TestConfiguration config = availableTestConfigurations.get(TEST_NAME);
+		loadTestConfiguration(config);
+
+		// Run reference dml script with normal matrix
+		fullDMLScriptName = HOME + TEST_NAME + "Reference.dml";
+		programArgs = new String[] {"-nvargs", input("X1"), input("X2"), expected("Z")};
+		runTest(null);
+
+		FEDInstructionUtils.fedDataGen = true;
+
+		// Run actual dml script with federated matrix
+		OptimizerUtils.FEDERATED_COMPILATION = true;
+		fullDMLScriptName = HOME + TEST_NAME + ".dml";
+
+		programArgs = new String[] {"-nvargs", "X1=" + TestUtils.federatedAddress(port1, input("X1")),
+			"X2=" + TestUtils.federatedAddress(port2, input("X2")),
+			"r=" + rows, "c=" + cols, "min=" + min, "max=" + max, "sp=" + sparsity, "seed=" + seed, "Z=" + output("Z")};
+		runTest(null);
+
+		HashMap<MatrixValue.CellIndex, Double> output = TestUtils.readDMLMatrixFromHDFS(outputDirectories[0]);
+		int rowOut = output.keySet().stream().max(Comparator.comparingInt(e -> e.row)).get().row;
+		int colOut = output.keySet().stream().max(Comparator.comparingInt(e -> e.column)).get().column;
+		double minOut = output.values().stream().min(Comparator.comparingDouble(Double::doubleValue)).get();
+		double maxOut = output.values().stream().max(Comparator.comparingDouble(Double::doubleValue)).get();
```

Comment:
this is not a nice comparison, since it only veifies max and min, and also steam through all cells
just make use of a simple compare of the outputs.